src/Common/OpenTelemetryTraceContext.h

```diff
@@ -1,24 +1,161 @@
+
+    /// Use weak_ptr instead of shared_ptr to hold a reference to the underlying system.opentelemetry_span_log table
+    /// Since this object is kept on threads and passed across threads, a weak_ptr is more safe to prevent potential leak
+    std::weak_ptr<OpenTelemetrySpanLog> span_log;
+};
+
+/// Holder of tracing context.
+/// It should be initialized at the beginning of each thread execution.
+/// And once it's destructed, it clears the context automatically.
+///
+/// It's also the root of all spans on current thread execution.
+///
+/// Although it's SAFE to construct this object multiple times on one same thread, it should be created at the beginning of one thread execution.
+struct TracingContextHolder
+{
+    /// Forbidden copy ctor and assignment to make the destructor safe
+    TracingContextHolder(const TracingContextHolder& scope) = delete;
+    TracingContextHolder& operator =(const TracingContextHolder& scope) = delete;
+
+    TracingContextHolder(std::string_view _operation_name,
+        const TracingContext& _parent_trace_context,
+        const std::weak_ptr<OpenTelemetrySpanLog>& _log)
+        : TracingContextHolder(_operation_name,
+            _parent_trace_context,
+            nullptr,
+            _log)
+    {
+    }
+
+    /// Initialize a tracing context on a child thread based on the context from the parent thread
+    TracingContextHolder(std::string_view _operation_name, const TracingContextOnThread & _parent_thread_trace_context)
+        : TracingContextHolder(_operation_name,
+            _parent_thread_trace_context,
+            nullptr,
+            _parent_thread_trace_context.span_log)
+    {
+    }
+
+    /// For servers like HTTP/TCP/GRPC to initialize tracing context on thread that process requests from clients
+    TracingContextHolder(std::string_view _operation_name,
+        TracingContext _parent_trace_context,
+        const Settings & _settings,
+        const std::weak_ptr<OpenTelemetrySpanLog> & _log)
+        : TracingContextHolder(_operation_name,
+            _parent_trace_context,
+            &_settings,
+            _log)
+    {
+    }
+
+    TracingContextHolder(std::string_view _operation_name,
+        TracingContext _parent_trace_context,
+        const Settings* settings_ptr,
+        const std::weak_ptr<OpenTelemetrySpanLog> & _log);
+
+    ~TracingContextHolder();
+
+    Span root_span;
+
+private:
+    bool is_context_owner = true;
+};
+
+using TracingContextHolderPtr = std::unique_ptr<TracingContextHolder>;
+
+/// A span holder that creates span automatically in a (function) scope if tracing is enabled.
+/// Once it's created or destructed, it automatically maitains the tracing context on the thread that it lives.
+struct SpanHolder : public Span
+{
+    SpanHolder(std::string_view);
+    ~SpanHolder();
+
+    /// Finish a span explicitly if needed.
+    /// It's safe to call it multiple times
+    void finish();
 };
 
 }
+
+using namespace OpenTelemetry;
```

Comment:
Thanks for moving the OpenTelemetry classes into a sub-namespace. Maybe you also want to remove this `using namespace` declaration from the header? It is often considered [not-so-great practice](https://rules.sonarsource.com/cpp/RSPEC-1003), and something like `OpenTelemetry::Span` will be less ambiguous than just `Span`.

Reply:
Updated